torchtune/modules/transformer.py

```diff
@@ -61,23 +103,126 @@ def forward(
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            mask (Optional[Tensor]): Optional boolean tensor which contains the attention mask
+                with shape [batch_size x seq_length x seq_length]. This is applied after
+                the query-key multiplication and before the softmax. A value of True in row i
+                and column j means token i attends to token j. A value of False means token i
+                does not attend to token j. If no mask is specified, a causal mask
+                is used by default. Default is None.
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
+            encoder_mask (Optional[Tensor]):  Cross attention boolean tensor with shape
+                [batch_size x x_seq_len x encoder_seq_len]
+            input_pos (Optional[Tensor]): Optional tensor which contains the position ids
+                of each token. During training, this is used to indicate the positions
+                of each token relative to its sample when packed, shape [b x s].
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
 
-        TODO:
-            - Make position of norm configurable
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
         """
+        # Skip cross attention when no secondary input
+        if encoder_input is None:
+            return x
+
+        # A mask of tokens (x) with no encoder_input
+        output_mask = self.output_mask(encoder_mask)
+
         # Input tensor and attention output have the same shape
         # [b, s, d]
         # Norm applied before self-attention
-        attn_out = self.attn(self.sa_norm(x), mask=mask, input_pos=input_pos)
+        attn_out = self.attn(
+            self.attn_norm(x), encoder_input, mask=encoder_mask, input_pos=input_pos
+        )
+        if output_mask is not None:
+            attn_out.masked_fill_(output_mask, 0)
 
         # Residual connection; shape: [batch_size, seq_length, embed_dim]
-        h = attn_out + x
+        h = self.attn_scale(attn_out) + x
```

Comment:
I commented this elsewhere, but why do we actually need to expose attn_scale in our transformer class? Shouldn't something like [this](https://gist.github.com/ebsmothers/7ed8a319b7566b76868299e245f5995b) work? (That example is for MLP, but should be easily extended to attention too). This isolates to a less-widely-used module and we can then keep the signature of higher-level components simple and put a lot of this on the builder.

Reply:
I agree that the gate can be moved into the attention module and the mlp. I prefer to keep the tanh gating module though as modules get shown in the model summary and it makes it easy to add in a builder.